Authentication as a github app and mocking upgrade

lachmanfrantisek opened the pull request:
- Improve saving responses.
    - Use Singleton for `PersistentObjectStorage`.
    - Enable saving responses by env-var (`RECORD_REQUESTS`).
- Add support for authentication as a github-app.


-----

### Usage:
- Set `RECORD_REQUESTS` env. variable to turn it on and import services directly from ogr.
    - Use `from ogr import PagureService`, not `from ogr.services.pagure import PagureService`.
    - You can use the following code in the `__init__.py` to enable mocking for whole module:
    ```python
    import os
    os.environ["RECORD_REQUESTS"] = "TRUE"
    ```
- Set the storage file name before the test:
    ```python
    PersistentObjectStorage().storage_file = persistent_data_file
    ```
    - In OGR we are using file per test -- we have the following code in the `setUp` in `TestCase`:
    ```python
        test_name = self.id() or "all"
        persistent_data_file = os.path.join(
            PERSISTENT_DATA_PREFIX, f"test_github_data_{test_name}.yaml"
        )
        PersistentObjectStorage().storage_file = persistent_data_file
    ```
- You can use `PersistentObjectStorage().is_write_mode` to check the mode.

-----

TODO:


- [x] less magic in ogr/__init__.py
- ~~[ ] docs~~ will make a new PR
- ~~[ ] fix the test for github-app~~ I am having some problems with python-crypto -- I can fix that later to not leave that here for so long.

jscotka commented:
I propose completely use env vars not just one to avoid have persistent storage in constructor  and parameter  is just used when set some env var.

In current implementation: setting persistent_storage parameter means that service class changes behaviour to mocking. (at least it is still in github and was in pagure)

Your proposed  implementation leads to, that you set persistent storage param, and this param is ignored without setting env var,
in one hand, this could be useful, that you can use same tests for direct testing and also for mocking, but could be confusing what this parameter mean, in that case at least would be good to raise some warning, that you are using persistent storage param without env var set.

lachmanfrantisek commented:
> Your proposed implementation leads to, that you set persistent storage param, and this param is ignored without setting env var,
> in one hand, this could be useful, that you can use same tests for direct testing and also for mocking, but could be confusing what this parameter mean, in that case at least would be good to raise some warning, that you are using persistent storage param without env var set.

It will probably fail if you use mock class and do not set the `persistent_storage`.

This is just a first working solution, I need to clean it a bit or use some different solution.

The problem is that the current solution (in master) works well but it is hard to understand for other people... We need something that will be easy to use for newcomers and also easy to use in the projects that use OGR.

TomasTomecek commented on ogr/services/mock/github.py:
```diff
@@ -0,0 +1,77 @@
+logger = logging.getLogger(__name__)
+
+
+# TODO: upstream this to PyGithub
```
already upstreamed

lachmanfrantisek replied:
That makes it much easier!

TomasTomecek commented:
You'll hate me for asking, but do we actually need mocking when we have the offline mode?

lachmanfrantisek commented:
> You'll hate me for asking, but do we actually need mocking when we have the offline mode?

Good question!

- Definitely, for OGR itself it is needed.
- Do we have offline mode already? I could miss something, I can see only the read-only mode.
- Otherwise, it depends. I still think it will be handy to have it. Like @jscotka pointed out, the following scenario would be useful:

```bash
export OGR_MOCK=1
export OGR_MOCK_FILE=response-file.yaml
packit propose-update
# attach the reponse-file to bug report
```

(It would be usefull not only for ogr but for all systems we need to comunicate with.)

lachmanfrantisek commented:
I've changed the ev_var to be more general and we can use it in other projects as well (e.g. packit). It will be less confusing for people that do not know OGR.

I've also added the `RESPONSE_FILE` variable used as a default for the response file.

TomasTomecek commented:
by offline mode, I meant the response recording

> > You'll hate me for asking, but do we actually need mocking when we have the offline mode?
> 
> Good question!
> 
>     * Definitely, for OGR itself it is needed.

is this still the case? do we really need the mock classes which replace real github/pagure classes?

>     * Otherwise, it depends. I still think it will be handy to have it. Like @jscotka pointed out, the following scenario would be useful:
> 
> 
> ```shell
> export MOCK_REQUESTS=1
> export RESPONSE_FILE=response-file.yaml
> packit propose-update
> # attach the response-file to the bug report
> ```

Theoretically, this could indeed be useful. On the flip side, we haven't encountered a single scenario where we would literally needed this - we can always reproduce ourselves.

TomasTomecek commented:
I finally read the code: we should invent a proper name for the response recording - it's not mocking, that's why I was confused.

lachmanfrantisek commented on ogr/__init__.py:
```diff
@@ -25,22 +25,77 @@
+
+mock_env = os.getenv("MOCK_REQUESTS")
+if mock_env:
+    import github
```
> what is reason of imports just inside `if`.

Yes, I would like to completely split the two cases -- mocking/non-mocking



> in ucase you use absolute imports, it should be okay and clean: eg `import ogr.services.mock.tweaks`
> 
> and move these new definitions `def new__requestEncode` there and here just replace it? to avoid bigger code in **init**.py?

I am not sure if I got you right, but definitely, I would like to clean the `__init__` a bit.

lachmanfrantisek replied:
I've moved as much as possible away. (Some parts cannot be moved because of the import dependencies...)
